The return type of a `comptime` expression isn't checked prior to its evaluation

### Zig Version

0.10.0-dev.3659+e5e6eb983

### Steps to Reproduce

Just do anything with `comptime`.

To make the results more pronounced, here's a sample `test.zig` file to illustrate the effect:
```zig
pub fn main() void {
    // The `u8` value that `foo` returns is discarded.
    // Should result in a swift compile error.
    comptime foo();
}

// Any function that takes time to evaluate would work for this test.
fn foo() u8 {
    const u32_max = @import("std").math.maxInt(u32);

    @setEvalBranchQuota(u32_max);
    var x: u32 = 0;
    while (x != u32_max) : (x +%= 1) {}

    return 0;
}
```

### Expected Behavior

The behavior should be the same without the `comptime` call.
```zig
pub fn main() void {
    // The `u8` value that `foo` returns is discarded.
    // Should result in a swift compile error.
    foo();
}

// omitted
```

Stage 2:
```sh-session
$ zig run test.zig
test.zig:4:8: error: value of type 'u8' ignored
    foo();
    ~~~^~
test.zig:4:8: note: all non-void values must be used
test.zig:4:8: note: this error can be suppressed by assigning the value to '_'
```

Stage 1:
```sh-session
$ zig run test.zig -fstage1
./test.zig:4:8: error: expression value is ignored
    foo();
       ^
```

### Actual Behavior

Rather than warning about a discarded value, the `comptime` expression is evaluated first.

This obviously takes time due to the nature of the loop and currently causes an OOM on my machine because of the issue described in https://github.com/ziglang/zig/issues/12493.

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -3254,17 +3254,16 @@ fn zirEnsureResultUsed(sema: *Sema, block: *Block, inst: Zir.Inst.Index) Compile
     const operand = try sema.resolveInst(inst_data.operand);
     const src = inst_data.src();
 
-    return sema.ensureResultUsed(block, operand, src);
+    return sema.ensureResultUsed(block, sema.typeOf(operand), src);
 }
 
 fn ensureResultUsed(
     sema: *Sema,
     block: *Block,
-    operand: Air.Inst.Ref,
+    ty: Type,
     src: LazySrcLoc,
 ) CompileError!void {
-    const operand_ty = sema.typeOf(operand);
-    switch (operand_ty.zigTypeTag()) {
+    switch (ty.zigTypeTag()) {
         .Void, .NoReturn => return,
         .ErrorSet, .ErrorUnion => {
             const msg = msg: {
@@ -3277,7 +3276,7 @@ fn ensureResultUsed(
         },
         else => {
             const msg = msg: {
-                const msg = try sema.errMsg(block, src, "value of type '{}' ignored", .{operand_ty.fmt(sema.mod)});
+                const msg = try sema.errMsg(block, src, "value of type '{}' ignored", .{ty.fmt(sema.mod)});
                 errdefer msg.destroy(sema.gpa);
                 try sema.errNote(block, src, msg, "all non-void values must be used", .{});
                 try sema.errNote(block, src, msg, "this error can be suppressed by assigning the value to '_'", .{});
@@ -6641,6 +6640,10 @@ fn analyzeCall(
                 };
             }
 
+            if (is_comptime_call and ensure_result_used) {
+                try sema.ensureResultUsed(block, fn_ret_ty, call_src);
+            }
+
             const result = result: {
                 sema.analyzeBody(&child_block, fn_info.body) catch |err| switch (err) {
                     error.ComptimeReturn => break :result inlining.comptime_result,
@@ -6763,7 +6766,7 @@ fn analyzeCall(
     };
 
     if (ensure_result_used) {
-        try sema.ensureResultUsed(block, result, call_src);
+        try sema.ensureResultUsed(block, sema.typeOf(result), call_src);
     }
     if (call_tag == .call_always_tail) {
         return sema.handleTailCall(block, call_src, func_ty, result);
@@ -7406,7 +7409,7 @@ fn instantiateGenericCall(
     sema.appendRefsAssumeCapacity(runtime_args);
 
     if (ensure_result_used) {
-        try sema.ensureResultUsed(block, result, call_src);
+        try sema.ensureResultUsed(block, sema.typeOf(result), call_src);
     }
     if (call_tag == .call_always_tail) {
         return sema.handleTailCall(block, call_src, func_ty, result);

--- a/test/cases/compile_errors/ignored_comptime_value.zig
+++ b/test/cases/compile_errors/ignored_comptime_value.zig
@@ -1,5 +1,17 @@
-export fn foo() void {
+export fn a() void {
     comptime 1;
+}
+export fn b() void {
+    comptime bar();
+}
+fn bar() u8 {
+    const u32_max = @import("std").math.maxInt(u32);
+
+    @setEvalBranchQuota(u32_max);
+    var x: u32 = 0;
+    while (x != u32_max) : (x +%= 1) {}
+
+    return 0;
 }
 
 // error
@@ -9,3 +21,6 @@ export fn foo() void {
 // :2:5: error: value of type 'comptime_int' ignored
 // :2:5: note: all non-void values must be used
 // :2:5: note: this error can be suppressed by assigning the value to '_'
+// :5:17: error: value of type 'u8' ignored
+// :5:17: note: all non-void values must be used
+// :5:17: note: this error can be suppressed by assigning the value to '_'

